### PR TITLE
Fix 1158

### PR DESCRIPTION
**Pull request by wind57**

### CODE REVIEW

File: spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointSliceV1CatalogWatch.java [hunk 1]
```diff
@@ -44,23 +45,25 @@
...
+
 		if (context.properties().allNamespaces()) {
 			LOG.debug(() -> "discovering endpoints in all namespaces");
-
-			// can't use try with resources here as it will close the client
-			KubernetesClient client = context.kubernetesClient();
 			endpointSlices = client.discovery().v1().endpointSlices().inAnyNamespace()
 					.withLabels(context.properties().serviceLabels()).list().getItems();
 		}
+		else if (!context.properties().namespaces().isEmpty()) {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointSliceV1CatalogWatch.java:55]: this is a core change, take namespaces into consideration, if such are present.

Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointSliceV1CatalogWatch.java:55]: if namespaces are present, take them, this is a core change

File: spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsCatalogWatch.java [hunk 2]
```diff
@@ -44,25 +45,25 @@
...
+
 		if (context.properties().allNamespaces()) {
 			LOG.debug(() -> "discovering endpoints in all namespaces");
-
-			// can't use try with resources here as it will close the client
-			KubernetesClient client = context.kubernetesClient();
 			endpoints = client.endpoints().inAnyNamespace().withLabels(context.properties().serviceLabels()).list()
 					.getItems();
 		}
+		else if (!context.properties().namespaces().isEmpty()) {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsCatalogWatch.java:56]: same thing here : take namespaces into consideration if they are set

File: spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/KubernetesCatalogWatch.java [hunk 3]
```diff
@@ -84,6 +84,12 @@ public void catalogServicesWatch() {
 
 	@PostConstruct
 	void postConstruct() {
+		stateGenerator = stateGenerator();
+	}
+
+	Function<Fabric8CatalogWatchContext, List<EndpointNameAndNamespace>> stateGenerator() {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/KubernetesCatalogWatch.java:90]: This eases the testing _a lot_, as I can create a Spy now from this class and only mock the result of this method.

File: spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * make sure that all the tests for endpoints are also handled by endpoint slices
+ *
+ * @author wind57
+ */
+abstract class Fabric8EndpointsAndEndpointSlicesTests {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java:57]: have a common abstract class that deals with all the boiler plate for both Endpoints and EndpointSlices. Also defines the methods that I want to test in both implementations.

File: spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+			boolean endpointSlices) {
+
+		when(NAMESPACE_PROVIDER.getNamespace()).thenReturn(namespace);
+
+		boolean allNamespaces = false;
+		KubernetesDiscoveryProperties properties = new KubernetesDiscoveryProperties(true, allNamespaces,
+				Set.of(namespace), true, 60, false, "", Set.of(), labels, "", null, 0, endpointSlices);
+		KubernetesCatalogWatch watch = new KubernetesCatalogWatch(mockClient(), properties, NAMESPACE_PROVIDER);
+
+		if (endpointSlices) {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java:265]: this is where I create a spy and mock just that method to return to me the EndpointSlices implementation.

File: spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+				.withTargetRef(new ObjectReferenceBuilder().withName(podName).withNamespace(namespace).build()).build();
+
+		EndpointSlice slice = new EndpointSliceBuilder().withMetadata(new ObjectMetaBuilder().withNamespace(namespace)
+				.withName("slice-" + podName).withLabels(labels).build()).withEndpoints(endpoint).build();
+
+		mockClient().discovery().v1().endpointSlices().inNamespace(namespace).resource(slice).create();
+
+	}
+
+	static void invokeAndAssert(KubernetesCatalogWatch watch, List<EndpointNameAndNamespace> state) {
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java:320]: all tests delated to this one, as it does the same thing

File: spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+		verify(APPLICATION_EVENT_PUBLISHER).publishEvent(HEARTBEAT_EVENT_ARGUMENT_CAPTOR.capture());
+
+		HeartbeatEvent event = HEARTBEAT_EVENT_ARGUMENT_CAPTOR.getValue();
+		assertThat(event.getValue()).isInstanceOf(List.class);
+
+		assertThat(event.getValue()).isEqualTo(state);
+	}
+
+	// work-around for : https://github.com/fabric8io/kubernetes-client/issues/4649
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java:331]: kind of un-expected from the fabric8 test extension... If I find time I will issue a PR against their repo also.

File: docs/src/main/asciidoc/discovery-client.adoc [hunk 8]
```diff
@@ -131,8 +131,23 @@ Spring Cloud Kubernetes can also watch the Kubernetes service catalog for change
 milliseconds (by default it is `30000`). The heartbeat event will contain the target references (and their namespaces of the addresses of all endpoints
 (for the exact details of what will get returned you can take a look inside `KubernetesCatalogWatch`). This is an implementation detail, and listeners of the heartbeat event
 should not rely on the details. Instead, they should see if there are differences between two subsequent heartbeats via `equals` method. We will take care to return a correct implementation that adheres to the equals contract.
-The endpoints will be queried in either all namespaces (enabled via `spring.cloud.kubernetes.discovery.all-namespaces=true`), or
-we will use: xref:property-source-config.adoc#namespace-resolution[Namespace Resolution].
+The endpoints will be queried in either :
```
Comment (wind57) [inline docs/src/main/asciidoc/discovery-client.adoc:134]: add documentation

**Comment by wind57** [comment]
@ryanjbaxter ready to be looked at when you get a chance. thank you

**Review by ryanjbaxter — APPROVED** [review]

**Pull request merged**